Include ARQ scheme #10

lukasostendorf opened the pull request:
This PR enhances the system with an additional MAC transmission mode.
Currently all MAC frames are sent in Unacknowledged mode, this PR allows to use a selective repeat ARQ in parallel to the existing Unacknowledged mode.

Fixes #10

lukasostendorf committed:
ARQ mode: change message definitions of dldata and uldata to contain
an Acknowledgement mode flag.

lukasostendorf committed:
add Unit Tests for some MAC messages

lukasostendorf committed:
fix linting

lukasostendorf committed:
ARQ: add constructor and parser for ACK messages

lukasostendorf committed:
ARQ: implement transmission window when sending fragments

lukasostendorf committed:
ARQ: include ACK transmissions

lukasostendorf committed:
ARQ: adapt reassembler to work with AM und UM mode

lukasostendorf committed:
ARQ: include handlers for ACK messages

lukasostendorf committed:
add Unit Tests for some MAC messages

lukasostendorf committed:
Bugfixes for ARQ:

- include packet inspection for UEs
- copy messages after adding them to the sender window. Otherwise
  messages get lost in window
- handle case in which there is data to send, but no fragment can be
  generated since the sender window is full

lukasostendorf committed:
ARQ: fix crash in mac fragmenter object

lukasostendorf committed:
ARQ: fix TCP/IP packet detection + add some logs

ethertype byte order was wrong
more log outputs to identify which ARQ scheme is currently used

lukasostendorf committed:
ARQ: bugfixes for fragmenter and reassembler

fragmenter_has_fragment returned true in some cases
were no fragment could be sent anymore

reassembler had a memory issue when reassembling a frame

lukasostendorf committed:
Mac messages: increase max sequence number for better flow with ARQ

we use 4 bits instead of 3 for seqNr

lukasostendorf committed:
ARQ: faster ACK transmission from Client to BS

BS will automatically assign UL slots when it transmits a frame that has
to be acked

lukasostendorf committed:
Fix mac uldata_ack and dldata_ack messages to 4Bit seqNr size

dl1com commented on src/mac/mac_messages.h:
```diff
@@ -40,18 +42,26 @@ typedef enum {
   mcs_chance_req,
+  dl_data_ack,
   ul_data = 15
 } CtrlID_e;
```
Please assign enum values explicitly to avoid potential confusion due to different enummeration when using different compilers.

lukasostendorf replied:
Fixed in https://github.com/HAMNET-Access-Protocol/HNAP4PlutoSDR/pull/37/commits/5ffd0417ecce7a9df6af43c2844005a180fcd570

dl1com commented on src/mac/mac_bs.c:
```diff
@@ -692,6 +709,19 @@ void *mac_bs_tap_rx_th(void *arg) {
       MacDataFrame frame = dataframe_create(dev->bytes_rec);
       memcpy(frame->data, dev->buffer, dev->bytes_rec);
 
+      // Packet inspection: determine if this is TCP traffic
```
Also used in mac_ue.c -> move this functionality to a common place?

lukasostendorf replied:
This is now moved to mac_common.c

dl1com commented on src/mac/mac_ue.c:
```diff
@@ -358,6 +374,19 @@ void *mac_ue_tap_rx_th(void *arg) {
     if (mac->tapdevice->bytes_rec > 0) {
       MacDataFrame frame = dataframe_create(mac->tapdevice->bytes_rec);
       memcpy(frame->data, mac->tapdevice->buffer, frame->size);
+      // Packet inspection: determine if this is TCP traffic
```
see comment in mac_bs.c

dl1com reviewed (commented):
I guess it would be nice to have some more verbose documentation on a per-function level, especially the input and output parameters. I'd propose to use doxygen format or something similar.
No need to do everything at once, but all functions which where touched by this PR may be a good start :)

lukasostendorf committed:
Client scheduler: check if a fragment is available before calling get_fragment

dl1com commented on src/mac/mac_ue.c:
```diff
@@ -250,8 +250,11 @@ void mac_ue_run_scheduler(MacUE mac) {
-          MacMessage msg = mac_frag_get_fragment(mac->fragmenter,
-                                                 lchan_unused_bytes(chan), 1);
+          MacMessage msg = NULL;
+          if (mac_frag_has_fragment(mac->fragmenter)) {
```
As there is no else criteria here, it might also make sense to include the call of the "mac_frag_has_fragment()" check into mac_frag_get_fragment()?
Otherwise it will probably forgotten to be called before other (future) usages of ...get_fragment() as well.

lukasostendorf replied:
This is a good idea, I've also included this in https://github.com/HAMNET-Access-Protocol/HNAP4PlutoSDR/pull/37/commits/5ffd0417ecce7a9df6af43c2844005a180fcd570

dl1com commented on src/mac/mac_messages.c:
```diff
@@ -409,8 +409,8 @@ void mac_msg_parse_mcs_change_req(MacMessage msg) {
-  msg->hdr.DLdataAck.seqNr = msg->hdr_bin[0] & 0b111;
+  msg->hdr.DLdataAck.ack_type = (msg->hdr_bin[0] >> 4) & 0b1;
+  msg->hdr.DLdataAck.seqNr = msg->hdr_bin[0] & 0b1111;
   msg->hdr.DLdataAck.fragNr = (msg->hdr_bin[1] >> 3) & 0b11111;
```
Is it possible to add a unit test for this?

dl1com commented on src/mac/mac_messages.c:
```diff
@@ -154,8 +154,8 @@ MacMessage mac_msg_create_ul_data_ack(uint8_t ack_type, uint8_t seqNr,
-  genericmsg->hdr_bin[0] |= (seqNr & 0b111);
+  genericmsg->hdr_bin[0] |= (ack_type & 0b1) << 4;
+  genericmsg->hdr_bin[0] |= (seqNr & 0b1111);
   genericmsg->hdr_bin[1] = (fragNr & 0b11111) << 3;
```
Is it possible to add a unit test for this?

lukasostendorf replied:
There is a unit test for dl_data_ack and ul_data_ack messages included in this branch.

test_mac_msg_parse_ul_data_ack()
test_mac_msg_parse_dl_data_ack()

in test_mac_messages.c

But I think it may be better to use the boundary values seqNr=15 and fragNr=31 for the test instead of the currently randomly chosen seqNr=5 and fragNr=13

dl1com commented on src/mac/mac_messages.c:
```diff
@@ -262,8 +262,8 @@ MacMessage mac_msg_create_dl_data_ack(uint8_t ack_type, uint8_t seqNr,
-  genericmsg->hdr_bin[0] |= (seqNr & 0b111);
+  genericmsg->hdr_bin[0] |= (ack_type & 0b1) << 4;
+  genericmsg->hdr_bin[0] |= (seqNr & 0b1111);
   genericmsg->hdr_bin[1] = (fragNr & 0b11111) << 3;
```
Is it possible to add a unit test for this?

dl1com commented on src/mac/mac_messages.c:
```diff
@@ -366,8 +366,8 @@ void mac_msg_parse_timing_advance(MacMessage msg) {
+  msg->hdr.ULdataAck.ack_type = (msg->hdr_bin[0] >> 4) & 0b1;
+  msg->hdr.ULdataAck.seqNr = msg->hdr_bin[0] & 0b1111;
   msg->hdr.ULdataAck.fragNr = (msg->hdr_bin[1] >> 3) & 0b11111;
 }
```
Is it possible to add a unit test for this?

lukasostendorf committed:
Adjust log levels for some PHY functions

lukasostendorf committed:
ARQ: clean up code

- include a maximum number of retransmits for the ARQ scheme
- check whether a fragment is available inside get_fragment
- Some doxygen documentation
- Verbose assignment of MacMessage types enum

dl1com reviewed (commented):
Nice work!
Just two more comments:

- Could you please add an unit test for `packet_inspect_is_tcpip`?
- Function `mac_frag_get_fragment` is of pretty high complexity (with over 100 lines) with a high level of indentations - I've opened an issue for refactoring: #42

lukasostendorf committed:
ARQ: refactor mac_frag_get_fragment for better readability

lukasostendorf committed:
ARQ: Bugfixes for mac_frag_get_fragment

lukasostendorf committed:
Add unit tests for packet inspection

lukasostendorf committed:
Fix linting

Commit:
Merge branch 'develop' into arq_scheme
